packages/protocol/engine/consensus/epochgadget/gadget.go

```diff
@@ -43,6 +43,14 @@ func (g *Gadget) LastConfirmedEpoch() epoch.Index {
 	return g.lastConfirmedEpoch
 }
 
+func (g *Gadget) setLastConfirmedEpoch(i epoch.Index) {
+	g.Lock()
+	defer g.Unlock()
+
+	g.lastConfirmedEpoch = i
+	g.Events.EpochConfirmed.Trigger(i)
```

Comment:
the trigger should maybe happen outside the scope of the lock to prevent possible deadlocks